File: backend/sentiment_sdk.py
"""
Sentiment SDK — Drop-in client for any Python project
======================================================
Copy this file into your project and use:

    from sentiment_sdk import Sentiment

    client = Sentiment(
        url='https://your-app.com',
        key='sk-sent-...',
    )

    result = client.analyze('Bitcoin is pumping!', source='twitter')
    print(result['label'])  # "POSITIVE"
"""

import requests
from typing import Optional


class Sentiment:
    def __init__(self, url: str, key: str):
        self.url = url.rstrip('/')
        self.key = key

    def _request(self, method: str, path: str, body: dict = None):
        resp = requests.request(
            method,
            f"{self.url}/api/v1/sentiment{path}",
            json=body,
            headers={
                "Content-Type": "application/json",
                "X-API-Key": self.key,
            },
            timeout=10,
        )
        data = resp.json()
        if not data.get("ok"):
            raise Exception(data.get("message") or data.get("error") or "API error")
        return data["data"]

    def analyze(self, text: str, source: Optional[str] = None) -> dict:
        """Analyze a single text."""
        return self._request("POST", "/analyze", {"text": text, "source": source})

    def batch(self, items: list, source: Optional[str] = None) -> dict:
        """Analyze multiple texts at once (up to 100)."""
        return self._request("POST", "/batch", {"items": items, "source": source})

    def normalize(self, text: str) -> dict:
        """Normalize text: cleanup, tokenize, detect language."""
        return self._request("POST", "/normalize", {"text": text})

    def health(self) -> dict:
        """Check engine health."""
        return self._request("GET", "/health")

File: backend/sentiment-sdk.ts
/**
 * Sentiment SDK — Drop-in client for any project
 * ================================================
 * Copy this file into your project and use:
 *
 *   import { Sentiment } from './sentiment-sdk'
 *
 *   const client = new Sentiment({
 *     url: 'https://your-app.com',
 *     key: 'sk-sent-...',
 *   })
 *
 *   const result = await client.analyze('Bitcoin is pumping!', 'twitter')
 *   console.log(result.label)  // "POSITIVE"
 */

interface SentimentConfig {
  url: string
  key: string
}

interface AnalyzeResult {
  label: 'POSITIVE' | 'NEUTRAL' | 'NEGATIVE'
  score: number
  source: string
  meta: {
    engineVersion: string
    confidence: string
    confidenceScore: number
    processingTimeMs: number
    cached: boolean
    detected: {
      positiveWords: string[]
      negativeWords: string[]
      neutralWords: string[]
    }
  }
}

interface NormalizeResult {
  original: string
  cleaned: string
  tokens: string[]
  lang: 'en' | 'ru' | 'ua' | 'unknown'
  charCount: number
  wordCount: number
}

interface BatchResult {
  results: Array<{
    id: string
    result: AnalyzeResult | null
    error: string | null
  }>
  meta: {
    totalItems: number
    successCount: number
    errorCount: number
    processingTimeMs: number
  }
}

export class Sentiment {
  private url: string
  private key: string

  constructor(config: SentimentConfig) {
    this.url = config.url.replace(/\/$/, '')
    this.key = config.key
  }

  private async request<T>(method: string, path: string, body?: any): Promise<T> {
    const res = await fetch(`${this.url}/api/v1/sentiment${path}`, {
      method,
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': this.key,
      },
      body: body ? JSON.stringify(body) : undefined,
    })

    const json = await res.json()
    if (!json.ok) throw new Error(json.message || json.error || 'API error')
    return json.data
  }

  /** Analyze a single text */
  async analyze(text: string, source?: string): Promise<AnalyzeResult> {
    return this.request('POST', '/analyze', { text, source })
  }

  /** Analyze multiple texts at once (up to 100) */
  async batch(items: Array<{ id: string; text: string; source?: string }>, source?: string): Promise<BatchResult> {
    return this.request('POST', '/batch', { items, source })
  }

  /** Normalize text: cleanup, tokenize, detect language */
  async normalize(text: string): Promise<NormalizeResult> {
    return this.request('POST', '/normalize', { text })
  }

  /** Check engine health */
  async health(): Promise<{ status: string; engineVersion: string }> {
    return this.request('GET', '/health')
  }
}
